_posts/2022-03-21-getting-started-with-protobufs-for-typescript.md

```diff
@@ -0,0 +1,259 @@
+
+Having got to the point where I understand enough to use them in anger, I now realise that trying to learn Protobufs for Typescript as a first step was probably the problem. There is almost nothing useful out there for learning Protobufs with Typescript. I actually had to revert to my first love, Python, before I could transfer that newly minted knowledge to TypeScript. As a result, I though I would put together this example for anyone else who is fighting along this path.
+
+## Other Stuff & Assumptions
+
+I am going to skip over the why Protobufs vs XML and the like, there's lots of that out there and if you are still reading this you probably have been told to just get on and make it work.
+
+I'm assuming you know: node.js, Typescript, npm and Jest. I am using Mac/Linux here (sorry Windows users but you can deal with your own world of pain). I will not be hand-holding you through the entire application build, but instead will be focusing on the concepts and things you'll need to know to to do the basics with Protobufs.
+
+# Making It Work
+We're going to work through an end to end process to demonstrate the key points to be able to do the basics with Protobufs and TypeScript.
+
+1. [Define the data structure for your messages](#define-the-data-structure-for-your-messages)
+2. [Compile the Typescript files you will use in your code](#compile-the-typescript-you-will-use-in-your-code)
+3. [Create a message](#create-a-message)
+4. [Serialise the message to Protobuf](#serialise-the-message-to-protobuf)
+5. [Do something with the serialised message](#do-something-with-the-serialised-message)
+6. [De-serialise the message](#de-serialise-the-message)
+
+## Define the Data Structure for Your Messages
+
+First thing to realise about Protobufs are they are just an efficient (small data size) way of encoding messages. Next thing to realise is that these messages have a pre-defined data structure. You need to know that data structure at both ends, to serialize the message before sending and then de-serialize the message back into something useable when it has been received. This is what the `.proto` files are for, they define the data structure of the messages in a language independent format.
```

Comment:
'are they are' is clunky, rewrite this sentence.

Reply:
"is they are" but going to cut this paragraph back